File: src/scripts/enqueue_task.ts
import path from "node:path";
import { randomUUID } from "node:crypto";

import { Command } from "commander";

import { SqliteStore } from "../db";
import { QueueError } from "../errors";
import { resolveLogPath } from "../logging";
import { RedisQueue } from "../queue";

const DEFAULT_DB_PATH = "/data/clawbrain/db/clawbrain.sqlite3";
const DEFAULT_REDIS_URL = "redis://127.0.0.1:6379/0";
const DEFAULT_QUEUE_NAME = "clawbrain:tasks";
const ALLOWED_AGENTS = [
  "TranslatorAgent",
  "PMAgent",
  "MobileAgent",
  "OCRAgent",
  "QAAgent",
  "CoderAgent",
  "BuilderAgent",
  "UXAgent",
  "DeployerAgent",
] as const;
const ALLOWED_JOB_TYPES = ["command", "codex"] as const;

interface Args {
  type: (typeof ALLOWED_JOB_TYPES)[number];
  repo: string;
  agent: (typeof ALLOWED_AGENTS)[number];
  command: string;
  requestText: string;
  sessionId: string | null;
  stack: string | null;
  dbPath: string;
  redisUrl: string;
  queueName: string;
}

function parseArgs(argv: string[]): Args {
  const program = new Command();
  program
    .option("--type <type>", "command|codex", "command")
    .requiredOption("--repo <repo>")
    .requiredOption("--agent <agent>")
    .option("--command <command>")
    .requiredOption("--request-text <text>")
    .option("--session-id <id>")
    .option("--stack <name>")
    .option("--db-path <path>", "DB path", process.env.CLAWBRAIN_DB_PATH ?? DEFAULT_DB_PATH)
    .option("--redis-url <url>", "Redis URL", process.env.CLAWBRAIN_REDIS_URL ?? DEFAULT_REDIS_URL)
    .option("--queue-name <name>", "Queue name", process.env.CLAWBRAIN_QUEUE_NAME ?? DEFAULT_QUEUE_NAME);

  program.parse(argv);
  const opts = program.opts<Record<string, string | undefined>>();

  const type = String(opts.type ?? "command").trim() as (typeof ALLOWED_JOB_TYPES)[number];
  const agent = String(opts.agent ?? "").trim() as (typeof ALLOWED_AGENTS)[number];

  if (!ALLOWED_JOB_TYPES.includes(type)) {
    throw new Error(`[FAIL] invalid --type '${type}'`);
  }
  if (!ALLOWED_AGENTS.includes(agent)) {
    throw new Error(`[FAIL] invalid --agent '${agent}'`);
  }

  return {
    type,
    repo: String(opts.repo ?? "").trim(),
    agent,
    command: String(opts.command ?? "").trim(),
    requestText: String(opts.requestText ?? "").trim(),
    sessionId: opts.sessionId ? String(opts.sessionId).trim() : null,
    stack: opts.stack ? String(opts.stack).trim() : null,
    dbPath: String(opts.dbPath ?? DEFAULT_DB_PATH),
    redisUrl: String(opts.redisUrl ?? DEFAULT_REDIS_URL),
    queueName: String(opts.queueName ?? DEFAULT_QUEUE_NAME),
  };
}

async function main(argv: string[]): Promise<number> {
  const args = parseArgs(argv);
  const taskId = randomUUID();
  const createdAt = new Date().toISOString();
  const dbPath = path.resolve(args.dbPath);

  let command = args.command;
  if (args.type === "command" && !command) {
    // eslint-disable-next-line no-console
    console.error("[FAIL] --command is required when --type=command");
    return 1;
  }

  if (args.type === "codex" && !command) {
    command = `codex exec --skip-git-repo-check --sandbox workspace-write -C /srv/projects/${args.repo} -`;
  }

  const logsDir = path.resolve(process.env.CLAWBRAIN_LOGS_DIR ?? "/data/clawbrain/logs");
  const logPath = resolveLogPath(taskId, logsDir);

  const store = new SqliteStore(dbPath, process.env.CLAWBRAIN_MIGRATIONS_DIR ?? path.resolve(process.cwd(), "migrations"));
  store.init();
  store.createTask({
    taskId,
    sessionId: args.sessionId,
    repo: args.repo,
    agent: args.agent,
    status: "queued",
    requestText: args.requestText,
    logPath,
    createdAt,
  });
  store.close();

  const queue = new RedisQueue({ redisUrl: args.redisUrl, queueName: args.queueName });
  try {
    await queue.ping();
    await queue.enqueue({
      task_id: taskId,
      session_id: args.sessionId,
      type: args.type,
      repo: args.repo,
      agent: args.agent,
      command,
      request_text: args.requestText,
      stack: args.stack,
      db_path: dbPath,
      created_at: createdAt,
    });
  } catch (error) {
    if (error instanceof QueueError) {
      // eslint-disable-next-line no-console
      console.error(`[FAIL] queue error: ${error.message}`);
    } else {
      // eslint-disable-next-line no-console
      console.error(`[FAIL] queue error: ${String(error)}`);
    }
    return 1;
  }

  // eslint-disable-next-line no-console
  console.log(
    JSON.stringify({
      task_id: taskId,
      type: args.type,
      queue_name: args.queueName,
      db_path: dbPath,
      log_path: logPath,
    }),
  );
  return 0;
}

if (require.main === module) {
  void main(process.argv).then((code) => process.exit(code));
}

File: scripts/enqueue_task.py
#!/usr/bin/env python3
"""CLI to enqueue runner tasks in Redis."""

from __future__ import annotations

import argparse
import json
import os
import sys
import uuid
from datetime import datetime, timezone
from pathlib import Path

# Ensure repo root is importable when running as: python scripts/enqueue_task.py
REPO_ROOT = Path(__file__).resolve().parent.parent
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

from runner.db import create_task
from runner.logging import resolve_log_path
from runner.queue import QueueError, RedisQueue

DEFAULT_DB_PATH = Path("/data/clawbrain/db/clawbrain.sqlite3")
DEFAULT_REDIS_URL = "redis://127.0.0.1:6379/0"
DEFAULT_QUEUE_NAME = "clawbrain:tasks"
ALLOWED_AGENTS = (
    "TranslatorAgent",
    "PMAgent",
    "MobileAgent",
    "OCRAgent",
    "QAAgent",
    "CoderAgent",
    "BuilderAgent",
    "UXAgent",
    "DeployerAgent",
)
ALLOWED_JOB_TYPES = ("command", "codex")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Enqueue a ClawBrain task.")
    parser.add_argument("--type", default="command", choices=ALLOWED_JOB_TYPES)
    parser.add_argument("--repo", required=True)
    parser.add_argument("--agent", required=True, choices=ALLOWED_AGENTS)
    parser.add_argument("--command")
    parser.add_argument("--request-text", required=True)
    parser.add_argument("--session-id")
    parser.add_argument("--stack")
    parser.add_argument(
        "--db-path",
        default=os.environ.get("CLAWBRAIN_DB_PATH", str(DEFAULT_DB_PATH)),
    )
    parser.add_argument(
        "--redis-url",
        default=os.environ.get("CLAWBRAIN_REDIS_URL", DEFAULT_REDIS_URL),
    )
    parser.add_argument(
        "--queue-name",
        default=os.environ.get("CLAWBRAIN_QUEUE_NAME", DEFAULT_QUEUE_NAME),
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    task_id = str(uuid.uuid4())
    created_at = datetime.now(timezone.utc).isoformat()
    db_path = Path(args.db_path).expanduser()
    job_type = args.type.strip()
    repo = args.repo.strip()
    agent = args.agent.strip()
    command = (args.command or "").strip()

    if job_type == "command" and not command:
        print("[FAIL] --command is required when --type=command")
        return 1

    if job_type == "codex" and not command:
        command = (
            f"codex exec --skip-git-repo-check --sandbox workspace-write "
            f"-C /srv/projects/{repo} -"
        )

    logs_dir = Path(os.environ.get("CLAWBRAIN_LOGS_DIR", "/data/clawbrain/logs")).resolve()
    log_path = resolve_log_path(task_id, logs_dir=logs_dir)

    payload = {
        "task_id": task_id,
        "session_id": args.session_id,
        "type": job_type,
        "repo": repo,
        "agent": agent,
        "command": command,
        "request_text": args.request_text,
        "stack": args.stack,
        "db_path": str(db_path),
        "created_at": created_at,
    }

    create_task(
        db_path,
        task_id=task_id,
        session_id=args.session_id,
        repo=repo,
        agent=agent,
        status="queued",
        request_text=args.request_text,
        log_path=str(log_path),
        created_at=created_at,
    )

    queue = RedisQueue(redis_url=args.redis_url, queue_name=args.queue_name)
    try:
        queue.ping()
        queue.enqueue(payload)
    except QueueError as exc:
        print(f"[FAIL] queue error: {exc}")
        return 1

    print(
        json.dumps(
            {
                "task_id": task_id,
                "type": job_type,
                "queue_name": args.queue_name,
                "db_path": str(db_path),
                "log_path": str(log_path),
            },
            ensure_ascii=True,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
